### ISSUE TITLE
NodeSelector may cause descheduler to abort when unnecessary

### ISSUE THREAD

**Issue by damemi** [kubernetes-sigs/descheduler#469]
**What version of descheduler are you using?**

descheduler version: latest

**Which descheduler CLI options are you using?**
--node-selector

**What did you do?**
When using `--node-selector` to restrict descheduler to just 1 node, [this section of code](https://github.com/kubernetes-sigs/descheduler/blob/4108362/pkg/descheduler/descheduler.go#L99-L110) prevents the descheduler from running with:

```
I1210 17:32:36.835916       1 descheduler.go:107] "The cluster size is 0 or 1 meaning eviction causes service disruption or degradation. So aborting.."
```

**What did you expect to see?**
Even if the cluster size is actually larger than 1, but we only want to restrict evictions to 1 node (such as only preventing duplicates from existing on 1 node) this failure still occurs. This seems to be due to `nodeSelector` being passed to the original node list that is used in this check.

**What did you see instead?**
I should still be able to run descheduler against a single node out of my cluster, because I may have other nodes available for evicted pods that will not cause degradation

**Comment by lixiang233** [comment]
Yes, descheduler should work normally if we got more than one node in cluster available even if we only selected 1 node. I can help fix this, and will submit PR recently.
/assign

**Comment by aveshagarwal** [comment]
node selector above is working as expected as it was initially designed to select smaller set of nodes from all nodes, and that means only those nodes are the intended nodes not only for eviction but also for where evicted pods will land. 
So when node selector selects only 1 node, that is like having 1-node cluster, and thats why it does not process further. 

To give some context, Descheduler was fundamentally designed to balance nodes in an imbalanced cluster, for example in case of LowNodeUtilization, it computes resources from all selected nodes, most likely all nodes, or could be smaller set of nodes as selected by node selector. This functionality with node selector was needed in multi-tenant scenarios for OpenShift.

If there is need to change the behavior of node selector, i.e., for having different behavior for different strategies,  that is fine, but looks like the above change could break LowNodeUtilization strategy.

**Comment by damemi** [comment]
@aveshagarwal that is a good point, that having a node selector over only 1 node would hurt the performance of LowNodeUtilization specifically. However, I believe the error that's logged should be updated to reflect that, because running other basic strategies against only 1 node in the cluster will not inherently cause service disruption if there are other nodes in the cluster.

Perhaps this error could be updated to a warning, and disable the LowNodeUtilization strategy with a clearer explanation why. It's an edge case for sure, but I think it is valid for users who know what they are doing to want to run certain other strategies against only 1 node.

**Comment by aveshagarwal** [comment]
My main concern is understanding and expectation around node selector behavior. It's not just about 1-node cluster which is just a special case.  Specifically, the comment in the original description "I should still be able to run Descheduler against a single node out of my cluster, because I may have other nodes available for evicted pods that will not cause degradation"  does not provide the right picture.

In general Descheduler works best, when it has whole view of the cluster it is working on that means all nodes. The node selector just causes Descheduler view to a smaller set of nodes as if the selected nodes are a cluster in themselves. That's why the expectation that evicted pods must land on only those nodes part of selected nodes.

When I had implemented node selector feature, the main design expectation for the node selector was that when a node selector is applied to a `x`-node cluster selecting `n` nodes (`x` >= `n`), it causes Descheduler to view only n nodes and x-n nodes are completely ignored. IOWs, node selector causes Descheduler's view reduced to only those nodes selected by the selector. 

The above is true for the right behavior of LowNodeUtilization.  

If there is a need to change the node selector behavior for other strategies, that's fine, but it would be better to document the expectations around node selectors and then make required implementation changes.

**Comment by lixiang233** [comment]
I thought that `Node Selector` is used to select nodes which descheduler can evicts pods from, and it doesn't matter whether these evicted pods will be rescheduled to selected nodes or not, but after @aveshagarwal  explained the initial design, it's more likely that we can only reschedule pods to other selected nodes(in descheduler's view, the cluster only contains selected nodes), so `descheduler` will abort if we only select 1 node.

BTW, if we want to change this design, these strategies should be disabled: `LowNodeUtilization`, `RemoveDuplicates`, `RemovePodsViolatingTopologySpreadConstraint`.

**Comment by damemi** [comment]
I think RemiveDuplicates should be fine, because that doesn't require multi-node calculations to be effective. For example if I have nodes A, B, C and only want to keep duplicates off node B then RemoveDuplicates should still work because it only checks for duplicate pods on that node (that's actually the use case that discovered this).

I agree with the other 2 though, because both of those require consideration of all nodes to effectively make eviction decisions.

**Comment by aveshagarwal** [comment]
Current behavior of node selector is in line with intended use case for RemoveDuplicates too. Node selector for RemoveDuplicates was not intended to remove from only some specific nodes. 

RemoveDuplicates evicts duplicates pods from nodes with the assumption there are other nodes that can schedule evicted pods. This strategy relies on kube scheduler's even spreading functionality too like LowNodeUtilization.

The initial use case for RemoveDuplicates was that if some nodes in a Kube cluster go down either due to maintenance or other failures causing some pods to reschedule on other nodes. However, when the failed or in-maintenance nodes or their replacements come back, Descheduler should help the duplicate pods to even spread again.

So for example, there are 4 nodes (either in whole cluster or selected nodes via node selector), an app that has 4 pods one on each node (only on those 4 nodes via node selector). If one of the nodes goes down, 3 nodes will have one pod on 2 nodes and 2 pods on one node. Now when the failed node comes back, Descheduler will evict the duplicate pod and it will get scheduled to the node (as this node will be part of nodes selected via node selector). 

I don't think that Descheduler can provide with very high probability the above behavior when pods are just selected from some specific nodes. It might work sometime and it might not work other times as the evicted pods can get rescheduled to the same node. It could also cause end users to use this strategy erroneously.  

@damemi  Could you explain you use case here to help us understand why node selector as it is now is not working as expected?

**Comment by damemi** [comment]
The use case came from a user who wanted to test the Descheduler on a single node before expanding it to their entire cluster. My confusion was due to the ambiguity of the error message, where it's not clear whether `service degradation` is referring to degradation of the cluster, or the descheduler (or both?)

What I am proposing is simply that the check for "cluster size" not necessarily be coupled to the node-selector flag, because that is not an accurate representation of the available nodes in the cluster. Descheduler could still consider the actual cluster size for its performance while restricting evictions to those on the selected node, similar to namespace filtering.

There could be use cases for other strategies as well. Say for security reasons you want to strictly enforce pod antiaffinity on a specific node, or that one node is frequently tainted for some reason. Those 2 strategies could make sense to run on a single node because even if there really are no other nodes in the cluster for them, you only care that they are not running on their current node. The code doesn't take any other nodes into consideration for these examples.

If hypotheticals like that are not within the expected behavior of node selector, then that's fine. What's important is that now this discussion is available to more clearly define the behavior of node selector. This could be an opportunity to update our documentation to explain our decisions around this.

@aveshagarwal thank you for shedding some light on the original design intentions. If you feel there's no change to be made here, then I trust your knowledge as the original author :)

**Comment by ingvagabund** [comment]
> should still be able to run descheduler against a single node out of my cluster, because I may have other nodes available for evicted pods that will not cause degradation

With --node-selector set, descheduler takes into account and evicts pods only from nodes targeted by the node selector. Though, a pod can get scheduled to nodes outside of the node selector.

> node selector above is working as expected as it was initially designed to select smaller set of nodes from all nodes, and that means only those nodes are the intended nodes not only for eviction but also for where evicted pods will land.

I disagree with `but also for where evicted pods will land` part. Descheduler does not guarantee where a pod gets rescheduled.

> If there is need to change the behavior of node selector, i.e., for having different behavior for different strategies, that is fine, but looks like the above change could break LowNodeUtilization strategy.

In the past we discussed a possibility of moving --node-selector option under each strategy as a new param. Descheduler already provides strategies which can run over non-empty set of nodes (including a singleton set). The original design goal to load-balance pods across nodes no longer applies to all strategies we have today in the descheduler.

Also as @damemi already [pointed out](https://github.com/kubernetes-sigs/descheduler/issues/469#issuecomment-743230265)

> My main concern is understanding and expectation around node selector behavior. It's not just about 1-node cluster which is just a special case. Specifically, the comment in the original description "I should still be able to run Descheduler against a single node out of my cluster, because I may have other nodes available for evicted pods that will not cause degradation" does not provide the right picture.

I second what @aveshagarwal wrote. The description of the issue needs to be more detailed.

> BTW, if we want to change this design, these strategies should be disabled: LowNodeUtilization, RemoveDuplicates, RemovePodsViolatingTopologySpreadConstraint.

I am against changing any behaviour around `--node-selector` option. Instead, I suggest to deprecate the option and have each strategy to handle the case. Disabling the mentioned strategies because there's only one node left can be surprising for a user. I'd rather have each strategy run and have it log a reason why it can not proceed at a moment.

> What I am proposing is simply that the check for "cluster size" not necessarily be coupled to the node-selector flag, because that is not an accurate representation of the available nodes in the cluster. Descheduler could still consider the actual cluster size for its performance while restricting evictions to those on the selected node, similar to namespace filtering.

Moving the check to each strategy is an acceptable solution given the argument of not coupling `--node-selector` with the check.

**Comment by lixiang233** [comment]
So, based on what was discussed above, we can move the parameter `nodeSelector` to `StrategyParameters` and the config file may look like:
```yaml
apiVersion: "descheduler/v1alpha1"
kind: "DeschedulerPolicy"
strategies:
  "PodLifeTime":
     enabled: true
     params:
        podLifeTime:
          maxPodLifeTimeSeconds: 86400
        nodeSelector: prod=dev
```
then, each strategy will be able to set and handle this parameter separately.

WDYT @damemi @aveshagarwal @ingvagabund

**Comment by ingvagabund** [comment]
Doing that, every strategy will need to have a node informer (one can run `nodeInformer.Lister().List(nodeSelector)` then).

```go
type strategyFunction func(
	ctx context.Context, client clientset.Interface,
	strategy api.DeschedulerStrategy,
	nodes []*v1.Node,
	podEvictor *evictions.PodEvictor,
)
```

->

```go
type strategyFunction func(
	ctx context.Context, client clientset.Interface,
	strategy api.DeschedulerStrategy,
	nodeInformer coreinformers.NodeInformer,
	podEvictor *evictions.PodEvictor,
)
```

Depending on how many nodes there are in a cluster, each strategy will spend more time on listing and filtering nodes.

**Comment by lixiang233** [comment]
@ingvagabund  I'm about to refactor [ReadyNodes](https://github.com/kubernetes-sigs/descheduler/blob/d502f059103ca87b69afbffa1f9bc736259a1971/pkg/descheduler/node/node.go#L33)  to return all ready nodes in cluster and descheduler will pass all of them to all strategies, then in every strategy we'll select nodes by its configured node selector, nodes will be listed only once and no strategy will need to have a node informer. Here's a code example:
```go
wait.Until(func() {
	nodes, err := nodeutil.ReadyNodes(ctx, rs.Client, nodeInformer)
	...
	for name, f := range strategyFuncs {
		if strategy := deschedulerPolicy.Strategies[api.StrategyName(name)]; strategy.Enabled {
			f(ctx, rs.Client, strategy, nodes, podEvictor)
		}
	}
}
...
func PodLifeTime() {
	selectedNodes, err := nodeutil.SelectNodes(nodes)
	...
}
```
Selecting nodes in each strategy will cost more time, but if we want to set `node selector` separately for each strategy, we must  bear the consequences.

**Comment by damemi** [comment]
I think how @lixiang233 described it is good, then we can just keep the one nodeinformer

**Comment by ingvagabund** [comment]
@lixiang233 iiuc what you are suggesting is the same as passing shared node informer to every strategy and running `nodeInformer.Lister().List(nodeSelector)` with subsequent is-node-ready filter.

**Comment by lixiang233** [comment]
@ingvagabund  yes, they're basically the same, what I'm suggesting only simplifies the function signature.

Selecting nodes will take time to traverse all nodes, but given nodes are stored in memory, I think the time it costs is acceptable.

**Comment by bwmetcalf** [comment]
Also, if running as a deployment, this issue causes the descheduler to enter `CrashLoopBackOff` with the following in the logs.  At the very least, why not just emit a message that the descheduler is not going to operate on one node and continue to run?
```
I0302 13:31:45.482881       1 node.go:46] "Node lister returned empty list, now fetch directly"
I0302 13:31:45.490349       1 descheduler.go:231] "The cluster size is 0 or 1 meaning eviction causes service disruption or degradation. So aborting.."
I0302 13:31:45.490439       1 reflector.go:225] Stopping reflector *v1.Pod (0s) from k8s.io/client-go/informers/factory.go:134
I0302 13:31:45.490480       1 reflector.go:225] Stopping reflector *v1.Namespace (0s) from k8s.io/client-go/informers/factory.go:134
I0302 13:31:45.490531       1 reflector.go:225] Stopping reflector *v1.PriorityClass (0s) from k8s.io/client-go/informers/factory.go:134
panic: close of closed channel

goroutine 77 [running]:
sigs.k8s.io/descheduler/pkg/descheduler.Run.func1()
	/go/src/sigs.k8s.io/descheduler/pkg/descheduler/descheduler.go:76 +0x3c
created by sigs.k8s.io/descheduler/pkg/descheduler.Run
	/go/src/sigs.k8s.io/descheduler/pkg/descheduler/descheduler.go:74 +0x165
```

**Comment by damemi** [comment]
@bwmetcalf could you provide an example of how you're using HighNodeUtilization to drain the single node? (I think I get your use case from the other discussion we had but just want to make sure I understand here)

For the panic, I tried to look at that line of code but it looks like that file has changed. What version of descheduler are you running? I agree that it shouldn't just panic and I don't think it did before, so that's something we should address if so

**Comment by bwmetcalf** [comment]
Hey @damemi.  Here is a link where I explained our use case in the context of karpenter.  Let me know if you need more information.  Thanks!  https://github.com/aws/karpenter/issues/1433#issuecomment-1057120579

**Comment by bwmetcalf** [comment]
Just back from vacation and didn't realize I didn't provide an answer to what version we are running.  It's 0.23.1, the latest.

**Comment by damemi** [comment]
Sorry I've been slow on this. Just checking, you still seeing that panic with the deployment? The panic looks like the one from https://github.com/kubernetes-sigs/descheduler/issues/728, which should have been fixed in `v0.23.1` by https://github.com/kubernetes-sigs/descheduler/pull/700. Also, that issue was related more to Job/CronJob than deployments, so it's weird to come up now. Maybe the fix didn't address something correlated with this nodeselector behavior

**Comment by bwmetcalf** [comment]
We are still running 0.23.1 and haven't changed anything, so I assume this is still an issue.  Also, we are running as a deployment.

**Comment by amannm** [comment]
yes, seems like `HighNodeUtilization` enabled while running a single node makes the descheduler process "abort" itself on startup

**Comment by kseniyashaydurova** [comment]
We also see similar issue for **PodLifeTime** strategy. We define below configuration for descheduler, it should choose 2 different nodes for its work with labels `apps=true,ops=true` in `deschedulerPolicy.nodeSelector`, but it sees 0 nodes, so the descheduler work is aborted by itself.

```
descheduler:
  chartVersion: 0.25.2
  enabled: true
  schedule: "0 0 * * *"
  cmdOptions:
    v: 3
  successfulJobsHistoryLimit: 1
  failedJobsHistoryLimit: 1
  deschedulerPolicy:
    nodeSelector: 'apps=true,ops=true'  <---------------------------------------------------------
    maxNoOfPodsToEvictPerNode: 10
    maxNoOfPodsToEvictPerNamespace: 10
    evictLocalStoragePods: true
    evictFailedBarePods: false
    ignorePvcPods: true
    strategies:
      RemoveDuplicates:
        enabled: false # disable if you run HPAs
        params:
          nodeFit: true
      RemovePodsHavingTooManyRestarts:
        enabled: true
        params:
          nodeFit: true
          podsHavingTooManyRestarts:
            podRestartThreshold: 20
            includingInitContainers: true
      RemovePodsViolatingNodeTaints:
        enabled: true
        params:
          nodeFit: true
      RemovePodsViolatingNodeAffinity:
        enabled: true
        params:
          nodeFit: true
          nodeAffinityType:
            - requiredDuringSchedulingIgnoredDuringExecution
      RemovePodsViolatingInterPodAntiAffinity:
        enabled: true
        params:
          nodeFit: true
      RemovePodsViolatingTopologySpreadConstraint:
        enabled: true
        params:
          nodeFit: true
          includeSoftConstraints: false
      LowNodeUtilization:
        enabled: false
        params:
          nodeFit: true
          nodeResourceUtilizationThresholds:
            thresholds:
              cpu: 40
              memory: 40
              # pods: 20
            targetThresholds:
              cpu: 70
              memory: 70
              # pods: 50
          namespaces:
            exclude:
              - prometheus
              - monitoring
              - ops
              - metrics-server

      PodLifeTime:                                 <---------------------------------------------------------
        enabled: true
        params:
          podLifeTime:
            maxPodLifeTimeSeconds: 86400 # 24h
          namespaces:
            include:
              - utils
              - teleport-agent

  nodeSelector: {}
  affinity: {}
  tolerations: {}
```

We get below logs:
```
I1226 08:59:08.349986       1 named_certificates.go:53] "Loaded SNI cert" index=0 certName="self-signed loopback" certDetail="\"apiserver-loopback-client@1672045148\" [serving] validServingFor=[apiserver-loopback-client] issuer=\"apiserver-loopback-client-ca@1672045147\" (2022-12-26 07:59:07 +0000 UTC to 2023-12-26 07:59:07 +0000 UTC (now=2022-12-26 08:59:08.349955936 +0000 UTC))"
I1226 08:59:08.350055       1 secure_serving.go:210] Serving securely on [::]:10258
I1226 08:59:08.350147       1 tlsconfig.go:240] "Starting DynamicServingCertificateController"
I1226 08:59:08.366628       1 reflector.go:221] Starting reflector *v1.PriorityClass (0s) from k8s.io/client-go/informers/factory.go:134
I1226 08:59:08.366644       1 reflector.go:257] Listing and watching *v1.PriorityClass from k8s.io/client-go/informers/factory.go:134
I1226 08:59:08.366680       1 reflector.go:221] Starting reflector *v1.Pod (0s) from k8s.io/client-go/informers/factory.go:134
I1226 08:59:08.366687       1 reflector.go:257] Listing and watching *v1.Pod from k8s.io/client-go/informers/factory.go:134
I1226 08:59:08.366863       1 reflector.go:221] Starting reflector *v1.Namespace (0s) from k8s.io/client-go/informers/factory.go:134
I1226 08:59:08.366870       1 reflector.go:257] Listing and watching *v1.Namespace from k8s.io/client-go/informers/factory.go:134
I1226 08:59:08.469186       1 node.go:49] "Node lister returned empty list, now fetch directly"
I1226 08:59:08.476678       1 descheduler.go:308] "The cluster size is 0 or 1 meaning eviction causes service disruption or degradation. So aborting.."
I1226 08:59:08.476832       1 reflector.go:227] Stopping reflector *v1.Pod (0s) from k8s.io/client-go/informers/factory.go:134
I1226 08:59:08.476900       1 reflector.go:227] Stopping reflector *v1.PriorityClass (0s) from k8s.io/client-go/informers/factory.go:134
I1226 08:59:08.476939       1 reflector.go:227] Stopping reflector *v1.Namespace (0s) from k8s.io/client-go/informers/factory.go:134
I1226 08:59:08.477106       1 tlsconfig.go:255] "Shutting down DynamicServingCertificateController"
I1226 08:59:08.477192       1 secure_serving.go:255] Stopped listening on [::]:10258
```

When I remove `deschedulerPolicy.nodeSelector` at all everything works ok and pods are killed by **PodLifeTime** strategy correctly.

**Comment by damemi** [comment]
@kseniyashaydurova I think your problem is that the nodeSelector combines all labels and only matches nodes that have all of those labels. This is why you're seeing 0 nodes listed.

From the [parser syntax](https://pkg.go.dev/k8s.io/apimachinery/pkg/labels#Parse) and [k8s docs on nodeSelector](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector) I don't think it's possible to set this up to match any of those labels, it needs to match all.

There are other options:
1. Apply a common label across your `apps` and `ops` nodes like `deschedulable=true`, then run descheduler against that label
2. Run 2 deschedulers (not recommended, and you might hit the 0/1 node problem again anyway)
3. Use a different label structure, for example a `role` label on your nodes ie `role=apps` and `role=ops`. This is a more descriptive use of node labels and allows you to run a single descheduler using the nodeSelector syntax like `nodeSelector: "role in (apps,ops)"`.

**Comment by a7i** [comment]
this was addressed in v0.23.1
feel free to reopen if that's not the case

/close
